### PR TITLE
Add support for auxclick event

### LINKED ISSUE #8529: Feature Request: Support auxclick event (onAuxClick)
Version 15.4.1

As starting Chrome 55, there is a new event `auxclick` to handle middle click, and `click` doesn't trigger by middle click anymore
https://developers.google.com/web/updates/2016/10/auxclick

### PR DESCRIPTION
**Pull request by jquense**
Adds support for `auxclick`. The little polyfill may be overkill, but it was simple enough to do so i put it in. Do we want to handle this or leave it up to users and add auxclick to the SimpleEvent plugin?

closes #8529

### CODE REVIEW

**Comment by nhunzaker** [comment]
Good question... How much do we want React to polyfill browser behavior? How do we decide what to polyfill?

**Comment by aweary** [comment]
Historically, we've avoided adding events that we couldn't polyfill, so IMO we definitely need to include this polyfill and make sure this works in unsupported browsers like IE.

**Comment by jquense** [comment]
A complicating factor with this event is that it was added along a side a change to click events, where they no longer fire for the middle mouse button. So even if it wasn't polyfillable i think there'd be a stronger case to add it (or use it to "fix" the click event)

File: packages/react-dom/src/events/AuxClickEventPlugin.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+const AuxClickEventPlugin = {
+  eventTypes,
+
+  extractEvents(
+    topLevelType: mixed,
+    targetInst: mixed,
+    nativeEvent: MouseEvent,
+    nativeEventTarget: EventTarget,
+  ) {
+    if (topLevelType === 'topClick' && nativeEvent.button === 0) {
```
Comment (nhunzaker) [inline packages/react-dom/src/events/AuxClickEventPlugin.js]: When is button not 0? Does this account for right clicks? Is that even a problem?

**Comment by jquense** [comment]
probably not worth the fixture? I'm not sure what it'd test that the unit test doesn't cover?

File: packages/react-dom/src/events/AuxClickEventPlugin.js (hunk 1 above)
Comment (jquense) [reply packages/react-dom/src/events/AuxClickEventPlugin.js]: this is any click that isn't a left click yeah. which is correct going off the mdn example https://mdn.github.io/dom-examples/auxclick/

**Comment by jquense** [comment]
one open question here: should the "polyfill" do something more advanced in the case where auxclick is support to prevent a case where click and auxclick both fire? I don't know if that will be the case ever, right now FF and Chrome both don't fire click anymore for none left clicks so it isn't an issue

**Comment by gaearon** [comment]
I'm very hesitant about adding more event polyfills given the focus on the bundle size.
What's the minimal constrained version we could do that would work only on supported browsers?

**Comment by jquense** [comment]
Minimimum is to add it the SimpleEventPlugin, and let the user handle the polyfill, e.g. listen for both `click` and `auxclick`

**Comment by gaearon** [comment]
That sounds okay as the first step to me.

File: packages/react-dom/src/events/SimpleEventPlugin.js [hunk 2]
```diff
@@ -136,7 +137,7 @@ const topLevelEventsToDispatchConfig: {
 });
 
 // Only used in DEV for exhaustiveness validation.
-const knownHTMLTopLevelTypes = [
+let knownHTMLTopLevelTypes = __DEV__ && [
```
Comment (nhunzaker) [inline packages/react-dom/src/events/SimpleEventPlugin.js]: Can this be reverted?

File: packages/react-dom/src/events/SimpleEventPlugin.js [hunk 3]
```diff
@@ -256,7 +259,7 @@ const SimpleEventPlugin: PluginModule<MouseEvent> = {
         break;
       default:
         if (__DEV__) {
-          if (knownHTMLTopLevelTypes.indexOf(topLevelType) === -1) {
+          if ((knownHTMLTopLevelTypes: any).indexOf(topLevelType) === -1) {
```
Comment (nhunzaker) [inline packages/react-dom/src/events/SimpleEventPlugin.js]: This too?

File: packages/react-dom/src/events/SimpleEventPlugin.js [hunk 4]
```diff
@@ -217,6 +219,7 @@ const SimpleEventPlugin: PluginModule<MouseEvent> = {
       case 'topMouseOut':
       case 'topMouseOver':
       case 'topContextMenu':
+
```
Comment (nhunzaker) [inline packages/react-dom/src/events/SimpleEventPlugin.js]: Could you nix this space?

**Comment by nhunzaker** [comment]
Upstreamed this with master. @gaearon do you mind taking a second look?

**Comment by nhunzaker** [comment]
@gaearon pushed a commit after some local testing. Just had to add aux click to the list of interactive event types (great change, @philipp-spiess btw)

**Comment by nhunzaker** [comment]
Waiting on CI for good measure, then I'll merge.

**Comment by gaearon** [comment]
Need to update snapshots.

**Comment by nhunzaker** [comment]
Huzzah.

**Comment by nhunzaker** [comment]
All set. Thanks @jquense!

**Comment by jquense** [comment]
thanks for cleaning this up @nhunzaker !

**Pull request merged**